File: LC/Python/Medium/54.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        m = len(matrix)
        n = len(matrix[0])
        
        i, j = 0, -1
        dirr = 1
        
        ans = []
        
        while 0 < m*n:
            for _ in range(n):
                j += dirr
                ans.append(matrix[i][j])
            m -= 1
            for _ in range(m):
                i += dirr
                ans.append(matrix[i][j])
            n -= 1
            dirr *= -1
            
        return ans

File: LC/C++/Medium/54.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        int m = matrix.size();
        int n = matrix[0].size();
        vector<int> ans;

        int i = 0, j = -1;
        int dirr = 1;

        while (0 < m*n) {
            for (int k = 0; k < n; ++k) {
                j += dirr;
                ans.emplace_back(matrix[i][j]);
            }
            --m;

            for (int k = 0; k < m; ++k){
                i += dirr;
                ans.emplace_back(matrix[i][j]);
            }
            --n;

            dirr *= -1;
        }

        return ans;
    }
};